java/client/src/org/openqa/selenium/virtualauthenticator/Credential.java

```diff
@@ -0,0 +1,116 @@
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.openqa.selenium.virtualauthenticator;
+
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * A credential stored in a virtual authenticator.
+ * @see https://w3c.github.io/webauthn/#credential-parameters
+ */
+public class Credential {
+
+  private final byte[] id;
+  private final boolean isResidentCredential;
+  private final String rpId;
+  private final PKCS8EncodedKeySpec privateKey;
+  private final byte[] userHandle;
+  private final int signCount;
+
+  /**
+   * Creates a non resident (i.e. stateless) credential.
+   */
+  public static Credential createNonResidentCredential(byte[] id, String rpId,
+      PKCS8EncodedKeySpec privateKey, int signCount) {
+    return new Credential(id, /*isResidentCredential=*/false, Objects.requireNonNull(rpId),
+        privateKey, /*userHandle=*/null, signCount);
+  }
+
+  /**
+   * Creates a resident (i.e. stateful) credential.
+   */
+  public static Credential createResidentCredential(byte[] id, String rpId,
+      PKCS8EncodedKeySpec privateKey, byte[] userHandle, int signCount) {
+    return new Credential(id, /*isResidentCredential=*/true, Objects.requireNonNull(rpId),
+        privateKey, Objects.requireNonNull(userHandle), signCount);
+  }
+
+  /**
+   * Creates a credential from a map.
+   */
+  public static Credential fromMap(Map<String, Object> map) {
+    Base64.Decoder decoder = Base64.getUrlDecoder();
+    return new Credential(decoder.decode((String) map.get("credentialId")),
+        (boolean) map.get("isResidentCredential"),
+        (String) map.get("rpId"),
+        new PKCS8EncodedKeySpec(decoder.decode((String) map.get("privateKey"))),
+        map.get("userHandle") == null ? null : decoder.decode((String) map.get("userHandle")),
+        ((Long) map.get("signCount")).intValue());
+  }
+
+  private Credential(byte[] id, boolean isResidentCredential, String rpId,
+      PKCS8EncodedKeySpec privateKey, byte[] userHandle, int signCount) {
+    this.id = Objects.requireNonNull(id);
+    this.isResidentCredential = isResidentCredential;
+    this.rpId = rpId;
+    this.privateKey = Objects.requireNonNull(privateKey);
+    this.userHandle = userHandle;
+    this.signCount = signCount;
+  }
+
+  public byte[] getId() {
+    return id;
+  }
+
+  public boolean getIsResidentCredential() {
+    return isResidentCredential;
+  }
+
+  public String getRpId() {
+    return rpId;
+  }
+
+  public PKCS8EncodedKeySpec getPrivateKey() {
+    return privateKey;
+  }
+
+  public byte[] getUserHandle() {
+    return userHandle;
+  }
+
+  public int getSignCount() {
+    return signCount;
+  }
+
+  public Map<String, Object> toMap() {
+    Base64.Encoder encoder = Base64.getUrlEncoder();
+    HashMap<String, Object> map = new HashMap();
```

Comment:
Nit: `Map` on the LHS. 

It doesn't matter a huge amount, but I'm curious why you've used `ImmutableMap` elsewhere and not here.

Reply:
I think an earlier version of my patch modified the map in place on `RemoteWebDriver`. Changed to `ImmutableMap` here and in `VirtualAuthenticatorOptions` for consistency. This means I had to add guava as a dependency to the bazel package -- ptal.

Reply:
Ah! We strive to make sure that the `selenium-api` doesn't need any third party libraries. If we want this to be immutable, the only Java 8 option is to use `Collections.unmodifiableMap`. The `Map.of()` method from Java 11 would be better, but we can't use that yet.

Reply:
Removed the dependency and used `Collections.unmodifiableMap` instead. Please take a look.